File: app/src/main/java/com/onemoresecret/KeyStoreListFragment.java
package com.onemoresecret;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.selection.ItemDetailsLookup;
import androidx.recyclerview.selection.ItemKeyProvider;
import androidx.recyclerview.selection.SelectionPredicates;
import androidx.recyclerview.selection.SelectionTracker;
import androidx.recyclerview.selection.StorageStrategy;
import androidx.recyclerview.widget.RecyclerView;

import com.onemoresecret.crypto.CryptographyManager;
import com.onemoresecret.databinding.FragmentKeyStoreListBinding;
import com.onemoresecret.databinding.PrivateKeyListItemBinding;

import java.security.KeyStoreException;
import java.security.interfaces.RSAPublicKey;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Enumeration;
import java.util.List;
import java.util.function.Consumer;

/**
 * A fragment representing a list of Items.
 */
public class KeyStoreListFragment extends Fragment {
    private FragmentKeyStoreListBinding binding;
    private SelectionTracker<String> selectionTracker;

    private final CryptographyManager cryptographyManager = new CryptographyManager();

    private final List<String> aliasList = new ArrayList<>();

    private final ItemAdapter itemAdapter = new ItemAdapter();

    private Consumer<FragmentKeyStoreListBinding> runOnStart;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentKeyStoreListBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    public SelectionTracker<String> getSelectionTracker() {
        return selectionTracker;
    }

    public void setRunOnStart(Consumer<FragmentKeyStoreListBinding> runOnStart) {
        this.runOnStart = runOnStart;
    }

    public void onItemRemoved(String alias) {
        int idx = aliasList.indexOf(alias);
        aliasList.remove(alias);
        //itemAdapter.notifyItemRemoved(idx); //this is not working
        itemAdapter.notifyDataSetChanged();
    }

    @Override
    public void onStart() {
        super.onStart();
        if (runOnStart != null) runOnStart.accept(binding);
        runOnStart = null;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        aliasList.clear();

        try {
            Enumeration<String> aliasesEnum = cryptographyManager.keyStore.aliases();
            while (aliasesEnum.hasMoreElements()) {
                aliasList.add(aliasesEnum.nextElement());
            }
            Collections.sort(aliasList);
        } catch (KeyStoreException e) {
            e.printStackTrace();
            throw new RuntimeException(e);
        }

        binding.list.setAdapter(itemAdapter);

        selectionTracker = new SelectionTracker.Builder<>("selectionTracker",
                binding.list,
                new PrivateKeyItemKeyProvider(ItemKeyProvider.SCOPE_MAPPED),
                new PrivateKeyLookup(),
                StorageStrategy.createStringStorage())
                .withSelectionPredicate(SelectionPredicates.createSelectSingleAnything())
                .build();
    }

    public OutputFragment getOutputFragment() {
        return (OutputFragment) binding.keyListOutputFragment.getFragment();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    class ItemAdapter extends RecyclerView.Adapter<PrivateKeyViewHolder> {

        @NonNull
        @Override
        public KeyStoreListFragment.PrivateKeyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            PrivateKeyListItemBinding binding =
                    PrivateKeyListItemBinding.inflate(LayoutInflater.from(parent.getContext()),
                            parent,
                            false);
            return new KeyStoreListFragment.PrivateKeyViewHolder(binding);
        }

        @Override
        public void onBindViewHolder(@NonNull KeyStoreListFragment.PrivateKeyViewHolder holder, int position) {
            holder.bind(position);
        }

        @Override
        public int getItemCount() {
            return aliasList.size();
        }

    }


    public class PrivateKeyViewHolder extends RecyclerView.ViewHolder {
        private final PrivateKeyListItemBinding binding;
        private String alias;
        private int position;

        public PrivateKeyViewHolder(@NonNull PrivateKeyListItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(int position) {
            this.position = position;
            this.alias = aliasList.get(position);
            try {
                binding.textItemKeyAlias.setText(alias);
                binding.textItemFingerprint.setText(
                        Util.byteArrayToHex(
                                CryptographyManager.getFingerprint(
                                        (RSAPublicKey) cryptographyManager.getCertificate(alias).getPublicKey())));
            } catch (Exception e) {
                e.printStackTrace();
            }
            binding.getRoot().setActivated(selectionTracker.isSelected(alias));
        }
    }


    public class PrivateKeyLookup extends ItemDetailsLookup<String> {

        @Nullable
        @Override
        public ItemDetails<String> getItemDetails(@NonNull MotionEvent e) {
            View view = binding.list.findChildViewUnder(e.getX(), e.getY());
            if (view == null)
                return null;

            PrivateKeyViewHolder holder = (PrivateKeyViewHolder) binding.list.getChildViewHolder(view);
            return new ItemDetails<String>() {
                @Override
                public int getPosition() {
                    return holder.position;
                }

                @Nullable
                @Override
                public String getSelectionKey() {
                    return holder.alias;
                }
            };
        }
    }


    public class PrivateKeyItemKeyProvider extends ItemKeyProvider<String> {

        /**
         * Creates a new provider with the given scope.
         *
         * @param scope Scope can't be changed at runtime.
         */
        protected PrivateKeyItemKeyProvider(int scope) {
            super(scope);
        }

        @Nullable
        @Override
        public String getKey(int position) {
            return aliasList.get(position);
        }

        @Override
        public int getPosition(@NonNull String key) {
            return aliasList.indexOf(key);
        }
    }

}

File: app/src/main/java/com/onemoresecret/KeyStoreListFragment.kt
package com.onemoresecret

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.fragment.app.Fragment
import androidx.recyclerview.selection.ItemDetailsLookup
import androidx.recyclerview.selection.ItemKeyProvider
import androidx.recyclerview.selection.SelectionPredicates
import androidx.recyclerview.selection.SelectionTracker
import androidx.recyclerview.selection.StorageStrategy
import androidx.recyclerview.widget.RecyclerView
import com.fasterxml.jackson.core.JsonProcessingException
import com.onemoresecret.composable.KeyStoreList
import com.onemoresecret.composable.OneMoreSecretTheme
import com.onemoresecret.composable.bindPrivateKeyListItem
import com.onemoresecret.composable.createPrivateKeyListItemComposeView
import com.onemoresecret.crypto.CryptographyManager
import com.onemoresecret.crypto.KeyStoreEntry
import com.onemoresecret.crypto.RSAUtil
import java.util.function.Consumer

class KeyStoreListFragment : Fragment() {
    private val aliasList = mutableListOf<String>()
    private val itemAdapter = ItemAdapter()
    private var keyStoreEntries = emptySet<KeyStoreEntry>()
    private var recyclerView: RecyclerView? = null
    private var tracker: SelectionTracker<String>? = null
    private var runOnStart: Consumer<SelectionTracker<String>>? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        loadKeyStoreEntries()

        return ComposeView(requireContext()).apply {
            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
            setContent {
                OneMoreSecretTheme {
                    KeyStoreList { view ->
                        if (recyclerView !== view) {
                            recyclerView = view
                            setupRecyclerView(view)
                        }
                    }
                }
            }
        }
    }

    fun getSelectionTracker(): SelectionTracker<String> {
        return requireNotNull(tracker)
    }

    fun setRunOnStart(runOnStart: Consumer<SelectionTracker<String>>) {
        this.runOnStart = runOnStart
        notifyRunOnStartIfReady()
    }

    fun onItemRemoved(alias: String) {
        val index = aliasList.indexOf(alias)
        if (index == -1) return
        aliasList.remove(alias)
        itemAdapter.notifyItemRemoved(index)
        //Update indices for items below the removed one
        itemAdapter.notifyItemRangeChanged(index, aliasList.size - index)
    }

    override fun onStart() {
        super.onStart()
        notifyRunOnStartIfReady()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        recyclerView = null
        tracker = null
    }

    private fun loadKeyStoreEntries() {
        aliasList.clear()
        val preferences = requireActivity().getPreferences(Context.MODE_PRIVATE)
        val keyStoreStringSet =
            preferences.getStringSet(CryptographyManager.PROP_KEYSTORE, HashSet()) ?: emptySet()

        keyStoreEntries = keyStoreStringSet.map { entry ->
            try {
                Util.JACKSON_MAPPER.readValue(entry, KeyStoreEntry::class.java)
            } catch (e: JsonProcessingException) {
                throw RuntimeException(e)
            }
        }.toSet()

        aliasList.addAll(keyStoreEntries.map { it.alias }.sorted())
    }

    private fun setupRecyclerView(view: RecyclerView) {
        view.adapter = itemAdapter
        tracker = SelectionTracker.Builder(
            "selectionTracker",
            view,
            PrivateKeyItemKeyProvider(ItemKeyProvider.SCOPE_MAPPED),
            PrivateKeyLookup(),
            StorageStrategy.createStringStorage()
        ).withSelectionPredicate(SelectionPredicates.createSelectSingleAnything()).build()
        notifyRunOnStartIfReady()
    }

    private fun notifyRunOnStartIfReady() {
        val selectionTracker = tracker ?: return
        runOnStart?.accept(selectionTracker)
        runOnStart = null
    }

    inner class ItemAdapter : RecyclerView.Adapter<PrivateKeyViewHolder>() {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PrivateKeyViewHolder {
            val composeView = createPrivateKeyListItemComposeView(parent)
            return PrivateKeyViewHolder(composeView)
        }

        override fun onBindViewHolder(holder: PrivateKeyViewHolder, position: Int) {
            holder.bind(position)
        }

        override fun getItemCount(): Int {
            return aliasList.size
        }
    }

    inner class PrivateKeyViewHolder(
        private val composeView: ComposeView
    ) : RecyclerView.ViewHolder(composeView) {
        var alias: String = ""
            private set
        var itemPosition: Int = RecyclerView.NO_POSITION
            private set

        fun bind(position: Int) {
            this.itemPosition = position
            alias = aliasList[position]
            try {
                val keyStoreEntry = keyStoreEntries.first { it.alias == alias }
                val publicKey = RSAUtil.restorePublicKey(keyStoreEntry.public)
                val fingerprint = Util.byteArrayToHex(RSAUtil.getFingerprint(publicKey))
                bindPrivateKeyListItem(
                    composeView,
                    alias,
                    fingerprint,
                    tracker?.isSelected(alias) == true
                )
            } catch (e: Exception) {
                throw RuntimeException(e)
            }

            composeView.isActivated = tracker?.isSelected(alias) == true
        }
    }

    inner class PrivateKeyLookup : ItemDetailsLookup<String>() {
        override fun getItemDetails(e: MotionEvent): ItemDetails<String>? {
            val list = recyclerView ?: return null
            val view = list.findChildViewUnder(e.x, e.y) ?: return null
            val holder = list.getChildViewHolder(view) as PrivateKeyViewHolder

            return object : ItemDetails<String>() {
                override fun getPosition(): Int {
                    return holder.itemPosition
                }

                override fun getSelectionKey(): String {
                    return holder.alias
                }
            }
        }
    }

    inner class PrivateKeyItemKeyProvider(scope: Int) : ItemKeyProvider<String>(scope) {
        override fun getKey(position: Int): String {
            return aliasList[position]
        }

        override fun getPosition(key: String): Int {
            return aliasList.indexOf(key)
        }
    }
}
